In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [5]:
query_gain_fiber = """
SELECT f4.block_group_code, f4.state, f4.county, f4.perc_urban, f4.population population14, f5.population population15, f6.population population16, f4.avg_n_of_tech70_servs AS service14, f5.avg_n_of_tech70_servs service15, f6.avg_n_of_tech70_servs service16
FROM broadband.fcc_acs_with_geo_and_props_2014 f4
     INNER JOIN broadband.fcc_acs_with_geo_and_props_2015 f5 ON f4.block_group_code = f5.block_group_code
     INNER JOIN broadband.fcc_acs_with_geo_and_props_2016 f6 ON f5.block_group_code = f6.block_group_code
WHERE (f4.avg_n_of_tech70_servs = 0 AND f5.avg_n_of_tech70_servs != 0) OR (f4.avg_n_of_tech70_servs = 0 AND f5.avg_n_of_tech70_servs = 0 AND f6.avg_n_of_tech70_servs != 0)
"""

query_gain_fiber_questionable = """
SELECT f4.block_group_code, f4.state, f4.county, f4.perc_urban, f4.population population14, f5.population population15, f6.population population16, f4.avg_n_of_tech70_servs AS service14, f5.avg_n_of_tech70_servs service15, f6.avg_n_of_tech70_servs service16 
FROM broadband.fcc_acs_with_geo_and_props_2014 f4
     INNER JOIN broadband.fcc_acs_with_geo_and_props_2015 f5 ON f4.block_group_code = f5.block_group_code
     INNER JOIN broadband.fcc_acs_with_geo_and_props_2016 f6 ON f5.block_group_code = f6.block_group_code
WHERE (f4.avg_n_of_tech70_servs = 0 AND f5.avg_n_of_tech70_servs != 0) OR (f5.avg_n_of_tech70_servs = 0 AND f6.avg_n_of_tech70_servs != 0)
"""

df_gain_fiber = pd.read_gbq(query_gain_fiber, project_id='fccdsicapstone-218522', dialect='standard')
df_gain_fiber_questionable = pd.read_gbq(query_gain_fiber_questionable, project_id='fccdsicapstone-218522', dialect='standard')

In [3]:
df_gain_fiber

block_group_code        state                   county  perc_urban  \
0         120050027042      Florida               Bay County    0.966102   
1         120050027043      Florida               Bay County    0.744444   
2         120050026071      Florida               Bay County    0.797619   
3         120050026051      Florida               Bay County    1.000000   
4         120050026042      Florida               Bay County    1.000000   
5         120050027051      Florida               Bay County    0.937500   
6         120050026062      Florida               Bay County    1.000000   
7         120050027022      Florida               Bay County    1.000000   
8         120050027023      Florida               Bay County    0.780000   
9         120050027053      Florida               Bay County    1.000000   
10        120050027032      Florida               Bay County    0.955556   
11        516400701014     Virginia               Galax city    0.978723   
12        516400701022     Virginia               Galax city    0.385965   
13        160459603005        Idaho               Gem County    0.264516   
14        160459601002        Idaho               Gem County    0.000000   
15        160459601003        Idaho               Gem County    0.307692   
16        160459602002        Idaho               Gem County    0.172414   
17        160459601001        Idaho               Gem County    0.002413   
18        160459601004        Idaho               Gem County    0.213333   
19        131770203002      Georgia               Lee County    0.232877   
20        131770201001      Georgia               Lee County    0.000000   
21        191114911003         Iowa               Lee County    0.446429   
22        191114907004         Iowa               Lee County    1.000000   
23        280819511004  Mississippi               Lee County    0.311475   
24        280819506022  Mississippi               Lee County    1.000000   
25        280819503023  Mississippi               Lee County    0.428571   
26        280819507003  Mississippi               Lee County    0.958904   
27        280819506021  Mississippi               Lee County    1.000000   
28        280819503022  Mississippi               Lee County    0.000000   
29        280819506011  Mississippi               Lee County    1.000000   
...                ...          ...                      ...         ...   
16401     220330042031    Louisiana  East Baton Rouge Parish    0.727273   
16402     220330045053    Louisiana  East Baton Rouge Parish    1.000000   
16403     220330039064    Louisiana  East Baton Rouge Parish    1.000000   
16404     220330023001    Louisiana  East Baton Rouge Parish    1.000000   
16405     220330036012    Louisiana  East Baton Rouge Parish    1.000000   
16406     220330042012    Louisiana  East Baton Rouge Parish    1.000000   
16407     220330004001    Louisiana  East Baton Rouge Parish    1.000000   
16408     220330040162    Louisiana  East Baton Rouge Parish    0.708333   
16409     220330017004    Louisiana  East Baton Rouge Parish    1.000000   
16410     220330046042    Louisiana  East Baton Rouge Parish    0.944444   
16411     220330045101    Louisiana  East Baton Rouge Parish    0.507463   
16412     220330035011    Louisiana  East Baton Rouge Parish    1.000000   
16413     220330039091    Louisiana  East Baton Rouge Parish    1.000000   
16414     220330034002    Louisiana  East Baton Rouge Parish    1.000000   
16415     220330046022    Louisiana  East Baton Rouge Parish    0.610169   
16416     220330037023    Louisiana  East Baton Rouge Parish    1.000000   
16417     220330016002    Louisiana  East Baton Rouge Parish    1.000000   
16418     220330040151    Louisiana  East Baton Rouge Parish    1.000000   
16419     220330031013    Louisiana  East Baton Rouge Parish    1.000000   
16420     220330031015    Louisiana  East Baton Rouge Parish    1.000000   
16421     220330040091    Louisiana  East Baton Rouge 

In [4]:
len(df_gain_fiber)

16431

In [6]:
len(df_gain_fiber_questionable)

18859

## What count as a block group code that gains a certain type of technology?

Take cable as an example. There are 16431

In [49]:
tech_code = ["10", "11", "12", "20", "30", "40", "41", "42", "43", "50", "70"]
tech_code = ["tech"+x for x in tech_code]

In [36]:
block_gain_fiber = """
SELECT f7.urban_rural
FROM 
      (SELECT *
      FROM broadband.fcc_allyear_blocklevel_with_geo_agg_2014
      WHERE file_year = 2014) f4
      RIGHT JOIN
      (SELECT *
      FROM broadband.fcc_allyear_blocklevel_with_geo_agg_2015
      WHERE file_year = 2015) f5 ON f4.block_code = f5.block_code
      RIGHT JOIN
      (SELECT *
      FROM broadband.fcc_allyear_blocklevel_with_geo_agg_2016
      WHERE file_year = 2016) f6 ON f5.block_code = f6.block_code 
      RIGHT JOIN
      (SELECT *
      FROM broadband.fcc_allyear_blocklevel_with_geo_agg_2017
      WHERE file_year = 2017) f7 ON f6.block_code = f7.block_code
WHERE (f4.n_of_[]_servs IS NULL AND f5.n_of_[]_servs > 0 AND f6.n_of_[]_servs > 0 AND f7.n_of_[]_servs > 0)
      OR (f4.n_of_[]_servs = 0 AND f5.n_of_[]_servs > 0 AND f6.n_of_[]_servs > 0 AND f7.n_of_[]_servs > 0)
      OR (f4.n_of_[]_servs IS NULL AND f5.n_of_[]_servs IS NULL AND f6.n_of_[]_servs > 0 AND f7.n_of_[]_servs > 0)
      OR (f4.n_of_[]_servs IS NULL AND f5.n_of_[]_servs = 0 AND f6.n_of_[]_servs > 0 AND f7.n_of_[]_servs > 0)
      OR (f4.n_of_[]_servs =0 AND f5.n_of_[]_servs = 0 AND f6.n_of_[]_servs > 0 AND f7.n_of_[]_servs > 0)
      OR (f4.n_of_[]_servs IS NULL AND f5.n_of_[]_servs IS NULL AND f6.n_of_[]_servs IS NULL AND f7.n_of_[]_servs > 0)
      OR (f4.n_of_[]_servs IS NULL AND f5.n_of_[]_servs IS NULL AND f6.n_of_[]_servs = 0 AND f7.n_of_[]_servs > 0)
      OR (f4.n_of_[]_servs IS NULL AND f5.n_of_[]_servs = 0 AND f6.n_of_[]_servs = 0 AND f7.n_of_[]_servs > 0)
      OR (f4.n_of_[]_servs = 0 AND f5.n_of_[]_servs = 0 AND f6.n_of_[]_servs = 0 AND f7.n_of_[]_servs > 0)
"""

2651097, 2186547

In [54]:
tech_code_dict_query = """
SELECT *
FROM broadband.tech_code
"""
tech_code_table = pd.read_gbq(tech_code_dict_query, project_id='fccdsicapstone-218522', dialect='standard')

In [71]:
tech_code_name = {}
for i in tech_code_table.iterrows():
    tech_code_name["tech" + str(i[1]["tech_code"])] = i[1]["tech_name"]

In [74]:
urban_rural = list()

for code in tech_code:
    df_block_gain_fiber = pd.read_gbq(block_gain_fiber.replace("[]", code), project_id='fccdsicapstone-218522', dialect='standard')
    ur_count = df_block_gain_fiber["urban_rural"].value_counts()
    urban_rural.append([code, tech_code_name[code],ur_count["U"], ur_count["R"], ur_count["U"]/(ur_count["U"]+ur_count["R"])])    

In [79]:
df_urban_rural = pd.DataFrame(urban_rural, columns = ["Tech_code", "Tech_name", "Urban", "Rural", "Perc_Urban"])

In [81]:
df_urban_rural.to_csv("block_gain_tech.csv")

In [28]:
type(df_block_gain_fiber["urban_rural"].value_counts())

pandas.core.series.Series

In [ ]:

block_gain_fiber = """
SELECT f7.block_code, f7.state, f7.county, f7.urban_rural, f7.urban_type, f7.total_length, f7.land_area, f7.elevation_mean, f4.n_of_[]_servs service14, f5.n_of_[]_servs service15, f6.n_of_[]_servs service16, f7.n_of_[]_servs service17
FROM 
      (SELECT *
      FROM broadband.fcc_allyear_blocklevel_with_geo_agg_2014
      WHERE file_year = 2014) f4
      RIGHT JOIN
      (SELECT *
      FROM broadband.fcc_allyear_blocklevel_with_geo_agg_2015
      WHERE file_year = 2015) f5 ON f4.block_code = f5.block_code
      RIGHT JOIN
      (SELECT *
      FROM broadband.fcc_allyear_blocklevel_with_geo_agg_2016
      WHERE file_year = 2016) f6 ON f5.block_code = f6.block_code 
      RIGHT JOIN
      (SELECT *
      FROM broadband.fcc_allyear_blocklevel_with_geo_agg_2017
      WHERE file_year = 2017) f7 ON f6.block_code = f7.block_code
WHERE (f4.n_of_[]_servs IS NULL AND f5.n_of_[]_servs > 0 AND f6.n_of_[]_servs > 0 AND f7.n_of_[]_servs > 0)
      OR (f4.n_of_[]_servs = 0 AND f5.n_of_[]_servs > 0 AND f6.n_of_[]_servs > 0 AND f7.n_of_[]_servs > 0)
      OR (f4.n_of_[]_servs IS NULL AND f5.n_of_[]_servs IS NULL AND f6.n_of_[]_servs > 0 AND f7.n_of_[]_servs > 0)
      OR (f4.n_of_[]_servs IS NULL AND f5.n_of_[]_servs = 0 AND f6.n_of_[]_servs > 0 AND f7.n_of_[]_servs > 0)
      OR (f4.n_of_[]_servs =0 AND f5.n_of_[]_servs = 0 AND f6.n_of_[]_servs > 0 AND f7.n_of_[]_servs > 0)
      OR (f4.n_of_[]_servs IS NULL AND f5.n_of_[]_servs IS NULL AND f6.n_of_[]_servs IS NULL AND f7.n_of_[]_servs > 0)
      OR (f4.n_of_[]_servs IS NULL AND f5.n_of_[]_servs IS NULL AND f6.n_of_[]_servs = 0 AND f7.n_of_[]_servs > 0)
      OR (f4.n_of_[]_servs IS NULL AND f5.n_of_[]_servs = 0 AND f6.n_of_[]_servs = 0 AND f7.n_of_[]_servs > 0)
      OR (f4.n_of_[]_servs = 0 AND f5.n_of_[]_servs = 0 AND f6.n_of_[]_servs = 0 AND f7.n_of_[]_servs > 0)
"""